### PR TITLE
[uxrce_dds_client] Set PX4 clock from UTC

### PR DESCRIPTION
**Pull request by dakejahl**
In systems with a companion computer without a GPS, it would be nice to use the timestamps from the DDS bridge to set the PX4 system time.

### CODE REVIEW

File: src/modules/uxrce_dds_client/uxrce_dds_client.cpp [hunk 1]
```diff
@@ -310,19 +310,43 @@ void UxrceddsClient::run()
...
+				// Check if clock is set
+				struct timespec ts = {};
+				px4_clock_gettime(CLOCK_REALTIME, &ts);
+				time_t current_utc = ts.tv_sec + (ts.tv_nsec / 1000000000);
 
-			} else {
-				usleep(10000);
+				if (current_utc > agent_utc) {
+					// Assume agents time is wrong if it's in the past
+					PX4_INFO("agents UTC time is %f seconds in the past, not setting clock",
```
Comment (dagar) [inline src/modules/uxrce_dds_client/uxrce_dds_client.cpp:327]:
```suggestion
					PX4_DEBUG("agents UTC time is %f seconds in the past, not setting clock",
```

**Comment by dagar** [comment]
This is likely what we want most of the time, but annoyingly not all the time.

At a minimum can you add some tolerance (so that we don't fight with GPS or Mavlink time setting) and probably best to add a parameter so that this can be disabled if necessary. 

https://github.com/PX4/PX4-Autopilot/blob/d6dbf38a1ba376cde9e65c10430e6feb52f6fdd5/src/drivers/gps/gps.cpp#L436-L441

Note for GPS right now that UTC time is coming from the navigation solution that's already ~100 milliseconds old. I have notes/plans on how to do it all perfectly with the TIMEPULSE, extending to CAN nodes, etc, but so far it hasn't been a priority.

**Comment by dakejahl** [comment]
> At a minimum can you add some tolerance (so that we don't fight with GPS or Mavlink time setting)

@dagar This logic occurs before the main loop and will only get executed once when the dds_client starts. Should I add it to the main loop such that it works the same way as GPS/Mavlink?

**Comment by Ryanf55** [comment]
Is there a reason you chose not to use MicroXRCE DDS's built in time sync support? 
https://micro-xrce-dds.docs.eprosima.com/en/latest/time_sync.html

I am happy to open up a discussion with eProsima to see if it's intended for this use case.

**Comment by dagar** [comment]
> > At a minimum can you add some tolerance (so that we don't fight with GPS or Mavlink time setting)
> 
> @dagar This logic occurs before the main loop and will only get executed once when the dds_client starts. Should I add it to the main loop such that it works the same way as GPS/Mavlink?

Likely yes, but we could come back to that later if/when we're ready to tackle the larger story with GPS (TIMEPULSE, etc). 

For now at least add the parameter so this can be disabled if it's not ideal for a particular setup?

**Comment by dakejahl** [comment]
@Ryanf55 it does use `uxr_set_time_callback`

@dagar okay I added a parameter to enable clock synchronization along with the same logic in GPS and Mavlink to only set the clock if `dt > 5_s`. I also refactored a little bit.. :sweat_smile:

**Comment by dakejahl** [comment]
Rebased and retested functional except for this issue which I think was introduced in the latest commits

**Comment by dakejahl** [comment]
@dagar asked about if this waits for the `Timesync` to converge -- the answer is "no", but neither does the current logic for setting the `session->time_offset`. Also I think it's really not important that the time is perfectly synchronized for this case anyway, we really just want to be close enough for the purpose of analyzing and organizing log files by time/date.

This PR stems from me trying to analyze logs from a system without GPS and all of the log files have the same garbage date which is the start of the epoch

**Comment by dagar** [comment]
All you'd need to do is make this public, then call `_timesync.sync_converged()`. We can do it later when tightening up the timesync in general (GPS, CAN nodes, etc).

https://github.com/PX4/PX4-Autopilot/blob/630dc993767078dae83d72c8823c474c805a65ac/src/lib/timesync/Timesync.hpp#L116-L120

**Review by dagar — APPROVED** [review]

**Pull request merged**